packages/data/src/index.js

```diff
@@ -307,25 +307,26 @@ export const withSelect = ( mapStateToProps ) => createHigherOrderComponent( ( W
 			this.unsubscribe();
 		}
 
-		shouldComponentUpdate( nextProps ) {
-			// This implementation of shouldComponentUpdate is only intended
-			// to handle incoming props changes. Changes effected via state
-			// changes are reflected by the subscribe handler's forceUpdate.
-			if ( isShallowEqual( this.props, nextProps ) ) {
+		shouldComponentUpdate( nextProps, nextState ) {
+			const hasPropsChanged = ! isShallowEqual( this.props, nextProps );
+
+			// Only render if result of props change or merge props update
+			// from store subscriber.
+			if ( this.state === nextState && ! hasPropsChanged ) {
 				return false;
 			}
 
-			// Regardless of whether merge props are changing, the component
-			// will need to render if props are different. But if the merge
-			// props change as a result of the incoming props, they should be
-			// reflected as such in the next render.
-			const nextMergeProps = getNextMergeProps( nextProps );
-			if ( ! isShallowEqual( this.mergeProps, nextMergeProps ) ) {
-				// Yes, this is a side effect. Yes, side effects don't belong
-				// in shouldComponentUpdate. This is not a typical component in
-				// that it is very much a hot code path, and will employ any
-				// and all gymnastics necessary for optimal performance.
-				this.mergeProps = nextMergeProps;
+			// If merge props change as a result of the incoming props, they
+			// should be reflected as such in the upcoming render.
+			if ( hasPropsChanged ) {
+				const nextMergeProps = getNextMergeProps( nextProps );
+				if ( ! isShallowEqual( this.mergeProps, nextMergeProps ) ) {
+					// While side effects are typically discouraged in this
+					// lifecycle methods, this component is very much on a hot
```

Comment:
"hot code path" is a great turn of phrase but it might confuse ESL readers (or just folks unfamiliar with that colloquium). I'm just gonna tweak this comment.